Add support for embeds_one and embeds_many

Currently we get the following error when using `embeds_one` and `embeds_many` i `JSONParser`. The attempted query was for a `get_wmbus_message_query/3` call. The code for the call can be found on the issue https://github.com/wise-home/wise_homex/pull/116

@lasseebert Will you take a look at it?

```
2020-01-09 13:26:35.893000 UTC [ERROR] [guest] <0.663.0> :gen_server: GenServer #PID<0.663.0> terminating
** (UndefinedFunctionError) function WiseHomex.Gateway.cast/1 is undefined or private
    (wise_homex) WiseHomex.Gateway.cast(nil)
    (wise_homex) lib/wise_homex/json_parser.ex:337: WiseHomex.JSONParser.parse_embed/2
    (wise_homex) lib/wise_homex/json_parser.ex:313: WiseHomex.JSONParser.convert_ecto_embedded_value/4
    (wise_homex) lib/wise_homex/json_parser.ex:229: WiseHomex.JSONParser.add_related_entities/2
    (wise_homex) lib/wise_homex/json_parser.ex:170: WiseHomex.JSONParser.add_relations/3
    (elixir) lib/enum.ex:1336: Enum."-map/2-lists^map/1-0-"/2
    (wise_homex) lib/wise_homex/json_parser.ex:64: WiseHomex.JSONParser.parse/1
    (wise_homex) lib/wise_homex/response_parser.ex:39: WiseHomex.ResponseParser.parse_json_response/2
    (arwen) lib/arwen_web/live/person_on_site_live.ex:69: ArwenWeb.PersonOnSiteLive.handle_info/2
    (phoenix_live_view) lib/phoenix_live_view/channel.ex:106: Phoenix.LiveView.Channel.handle_info/2
    (stdlib) gen_server.erl:637: :gen_server.try_dispatch/4
    (stdlib) gen_server.erl:711: :gen_server.handle_msg/6
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
```

--- a/lib/wise_homex/api_config.exs
+++ b/lib/wise_homex/api_config.exs
@@ -8,5 +8,13 @@ endpoints = [
     name_singular: "expense",
     path: "/expenses",
     type: "expenses"
+  },
+  %{
+    endpoints: [:create],
+    model: WiseHomex.WMBusMessageQuery,
+    name_plural: "wmbus_message_queries",
+    name_singular: "wmbus_message_query",
+    path: "/wmbus-message-queries",
+    type: "wmbus-message-queries"
   }
 ]

--- a/lib/wise_homex/json_parser.ex
+++ b/lib/wise_homex/json_parser.ex
@@ -44,7 +44,7 @@ defmodule WiseHomex.JSONParser do
              "utility-readings" => WiseHomex.UtilityReading,
              "wmbus-device-types" => WiseHomex.WMBusDeviceType,
              "wmbus-manufacturers" => WiseHomex.WMBusManufacturer,
-             "wmbus-message-infos" => WiseHomex.WMBusDeviceInfo,
+             "wmbus-message-infos" => WiseHomex.WMBusMessageInfo,
              "zip-codes" => WiseHomex.ZipCode
            }
            |> Map.merge(WiseHomex.ApiDefinition.type_to_model_mappings())

--- a/lib/wise_homex/json_parser.ex
+++ b/lib/wise_homex/json_parser.ex
@@ -44,6 +44,7 @@ defmodule WiseHomex.JSONParser do
              "utility-readings" => WiseHomex.UtilityReading,
              "wmbus-device-types" => WiseHomex.WMBusDeviceType,
              "wmbus-manufacturers" => WiseHomex.WMBusManufacturer,
+             "wmbus-message-infos" => WiseHomex.WMBusDeviceInfo,
              "zip-codes" => WiseHomex.ZipCode
            }
            |> Map.merge(WiseHomex.ApiDefinition.type_to_model_mappings())

--- a/lib/wise_homex/models/wmbus_message_info.ex
+++ b/lib/wise_homex/models/wmbus_message_info.ex
@@ -1,0 +1,14 @@
+defmodule WiseHomex.WMBusMessageInfo do
+  @moduledoc """
+  A WMBus Message Info
+  """
+
+  use WiseHomex.BaseModel
+
+  embedded_schema do
+    embeds_one :device, WiseHomex.Device
+
+    field :signal_strength, :integer
+    field :time, :utc_datetime
+  end
+end

--- a/lib/wise_homex/models/wmbus_message_info.ex
+++ b/lib/wise_homex/models/wmbus_message_info.ex
@@ -6,7 +6,7 @@ defmodule WiseHomex.WMBusMessageInfo do
   use WiseHomex.BaseModel
 
   embedded_schema do
-    embeds_one :device, WiseHomex.Device
+    belongs_to :device, WiseHomex.Device
 
     field :signal_strength, :integer
     field :time, :utc_datetime

--- a/lib/wise_homex/models/wmbus_message_query.ex
+++ b/lib/wise_homex/models/wmbus_message_query.ex
@@ -8,7 +8,7 @@ defmodule WiseHomex.WMBusMessageQuery do
   embedded_schema do
     embeds_many :devices, WiseHomex.Device
     embeds_many :wmbus_message_infos, WiseHomex.WMBusMessageInfo
-    embeds_one :gateway, WiseHomex.Gateway
+    belongs_to :gateway, WiseHomex.Gateway
 
     field :save, :boolean
     field :timeout, :integer

--- a/lib/wise_homex/models/wmbus_message_query.ex
+++ b/lib/wise_homex/models/wmbus_message_query.ex
@@ -1,0 +1,16 @@
+defmodule WiseHomex.WMBusMessageQuery do
+  @moduledoc """
+  A WMBus Message Query
+  """
+
+  use WiseHomex.BaseModel
+
+  embedded_schema do
+    embeds_many :devices, WiseHomex.Device
+    embeds_many :wmbus_message_infos, WiseHomex.WMBusMessageInfo
+    embeds_one :gateway, WiseHomex.Gateway
+
+    field :save, :boolean
+    field :timeout, :integer
+  end
+end

--- a/test/json_parser_test.exs
+++ b/test/json_parser_test.exs
@@ -891,6 +891,26 @@ defmodule WiseHomex.JSONParserTest do
     wmbus_message_query = JSONParser.parse(data)
     assert wmbus_message_query.devices |> length() == 2
     assert wmbus_message_query.gateway
+    assert wmbus_message_query.gateway_id == "2389"
     assert wmbus_message_query.wmbus_message_infos |> length() == 2
+  end
+
+  test "parse embeds that are not included" do
+    data = %{
+      "data" => %{
+        "attributes" => %{"save" => true, "timeout" => 30_000},
+        "id" => "",
+        "relationships" => %{
+          "gateway" => %{"data" => %{"id" => "7", "type" => "gateways"}}
+        },
+        "type" => "wmbus-message-queries"
+      },
+      "jsonapi" => %{"version" => "1.0"}
+    }
+
+    wmbus_message_query = JSONParser.parse(data)
+    assert wmbus_message_query.devices |> length() == 0
+    assert wmbus_message_query.gateway_id == "7"
+    assert wmbus_message_query.wmbus_message_infos |> length() == 0
   end
 end

--- a/test/json_parser_test.exs
+++ b/test/json_parser_test.exs
@@ -788,4 +788,109 @@ defmodule WiseHomex.JSONParserTest do
     heat_source = JSONParser.parse(data)
     assert heat_source.reduction_factor == ~d[1.10]
   end
+
+  test "parse embeds that are normal ecto schemas" do
+    data = %{
+      "data" => %{
+        "attributes" => %{"save" => true, "timeout" => 30_000},
+        "id" => "",
+        "relationships" => %{
+          "devices" => %{"data" => [%{"id" => "8350", "type" => "devices"}, %{"id" => "8351", "type" => "devices"}]},
+          "gateway" => %{"data" => %{"id" => "2389", "type" => "gateways"}},
+          "wmbus-message-infos" => %{
+            "data" => [
+              %{"id" => "286dbd35-93da-4926-b927-5a2202b9e350", "type" => "wmbus-message-infos"},
+              %{"id" => "a961b901-092e-4bb0-9c35-5a04fef2c888", "type" => "wmbus-message-infos"}
+            ]
+          }
+        },
+        "type" => "wmbus-message-queries"
+      },
+      "included" => [
+        %{
+          "attributes" => %{
+            "address" => nil,
+            "enabled-modules" => [],
+            "last-connect-at" => "2020-01-09T13:06:14Z",
+            "location" => nil,
+            "notes" => nil,
+            "online" => "no",
+            "postal-address" => nil,
+            "serial" => "0200000100003A7C",
+            "skip-offline-report" => false,
+            "software-version" => "balrog 1.2.3",
+            "unlocked-at" => "2020-01-09T13:06:14Z",
+            "unlocked-seconds" => 60
+          },
+          "id" => "2389",
+          "relationships" => %{"admin" => %{"data" => nil}, "sim" => %{"data" => nil}},
+          "type" => "gateways"
+        },
+        %{
+          "attributes" => %{
+            "authorized-at" => nil,
+            "inserted-at" => "2020-01-09T13:06:14Z",
+            "installation-year" => 2019,
+            "last-seen" => nil,
+            "number" => "00000107",
+            "online" => "no",
+            "protocol" => "wmbus",
+            "reduction-factor" => "1",
+            "serial" => "070100002D2C3004",
+            "signal-strength" => nil,
+            "signal-strength-history" => []
+          },
+          "id" => "8350",
+          "relationships" => %{
+            "device-type" => %{"data" => %{"id" => "4", "type" => "device-types"}},
+            "gateway" => %{"data" => nil},
+            "heat-source" => %{"data" => nil},
+            "room" => %{"data" => nil}
+          },
+          "type" => "devices"
+        },
+        %{
+          "attributes" => %{
+            "authorized-at" => nil,
+            "inserted-at" => "2020-01-09T13:06:14Z",
+            "installation-year" => 2019,
+            "last-seen" => nil,
+            "number" => "00000108",
+            "online" => "no",
+            "protocol" => "wmbus",
+            "reduction-factor" => "1",
+            "serial" => "080100002D2C3004",
+            "signal-strength" => nil,
+            "signal-strength-history" => []
+          },
+          "id" => "8351",
+          "relationships" => %{
+            "device-type" => %{"data" => %{"id" => "4", "type" => "device-types"}},
+            "gateway" => %{"data" => nil},
+            "heat-source" => %{"data" => nil},
+            "room" => %{"data" => nil}
+          },
+          "type" => "devices"
+        },
+        %{
+          "attributes" => %{"signal-strength" => -40, "time" => "2020-01-09T13:06:14.492626Z"},
+          "id" => "286dbd35-93da-4926-b927-5a2202b9e350",
+          "relationships" => %{"device" => %{"data" => %{"id" => "8350", "type" => "devices"}}},
+          "type" => "wmbus-message-infos"
+        },
+        %{
+          "attributes" => %{"signal-strength" => -40, "time" => "2020-01-09T13:06:14.493500Z"},
+          "id" => "a961b901-092e-4bb0-9c35-5a04fef2c888",
+          "relationships" => %{"device" => %{"data" => %{"id" => "8351", "type" => "devices"}}},
+          "type" => "wmbus-message-infos"
+        }
+      ],
+      "jsonapi" => %{"version" => "1.0"}
+    }
+
+    wmbus_message_query = JSONParser.parse(data)
+    assert wmbus_message_query.devices |> length() == 2
+    assert wmbus_message_query.gateway
+    assert wmbus_message_query.wmbus_message_infos |> length() == 2
+  end
 end

